File: Funk/lexer.py
from .utils.tokens import *

class Lexer:
  def __init__(self, source_code):
    self.source_code: str = source_code
    self.current_char: str = None
    self.pos: int = -1
    self.line: int = 1
    self.line_pos: int = 0
    self.tokens: list[Token] = []

  def next_char(self):
    if not self.source_code:
      self.current_char = None
      return
    self.current_char, self.source_code = self.source_code[0], self.source_code[1:]
    self.pos += 1
    self.line_pos += 1

  def lex(self):
    self.next_char()
    while self.current_char:
      match self.current_char:
        case c if c.isspace():
          if c == "\n":
            self.tokens.append(Token(TokenType.NEWLINE, "\n"))
            self.line_pos = 0
            self.line += 1
            self.next_char()
            continue
          self.tokens.append(Token(TokenType.WHITESPACE, " "))
          self.next_char()
        case c if c.isnumeric():
          self.tokens.append(self.handle_numbers())
        case c if c == "\"":
          self.tokens.append(self.handle_strings())
        case c if c in OPERATORS:
          self.tokens.append(self.handle_operators())
        case c if c.startswith("_") or c.isalpha():
          self.tokens.append(self.handle_keywords_and_variables())
        case c if c == "(":
          self.tokens.append(Token(TokenType.LPar, self.current_char))
          self.next_char()
        case c if c == ")":
          self.tokens.append(Token(TokenType.RPar, self.current_char))
          self.next_char()
        case _:
          # TODO: Make an Exception called UnknownCharacter and raise it
          self.next_char()

  def handle_numbers(self) -> Token:
    value = ""
    while self.current_char and self.current_char.isnumeric() or self.current_char == ".":
      value += self.current_char
      self.next_char()

    return Token(TokenType.NUMBER, value)

  def handle_strings(self) -> Token:
    value = ""
    self.next_char()
    while self.current_char and self.current_char != "\"":
      value += self.current_char
      self.next_char()
    if self.current_char != "\"":
      raise Exception("String was never closed.")
    self.next_char()
    return Token(TokenType.STRING, value)

  def handle_operators(self) -> Token:
    value = ""
    while self.current_char and self.current_char in OPERATORS:
      value += self.current_char
      self.next_char()
    return Token(TokenType.OPERATOR, value)
  
  def handle_keywords_and_variables(self) -> Token:
    value = self.current_char
    self.next_char()
    while self.current_char and (self.current_char.isalnum() or self.current_char == "_"):
      value += self.current_char
      self.next_char()
    if value in KEYWORDS:
      return Token(TokenType.KEYWORD, value)
    return Token(TokenType.IDENTIFIER, value)

File: src/lexer.rs
use std::slice::Iter;
use std::iter::Peekable;
use std::ops::Range;
use structs::structs::*;

pub struct Lexer {
  pub code: String,
  pub pos: usize,
  pub last_pos: usize,
  pub line_pos: usize,
  pub line: usize,
  pub current_char: Option<char>,
  pub next_char: Option<char>
}

impl Lexer {
  pub fn lex(&mut self) -> Vec<Token> {
    let chars = self.code.chars().collect::<Vec<char>>();

    self.last_pos = chars.len() - 1;
    let mut iter = chars.iter().peekable();

    self.parse_tokens(&mut iter)
  }

  fn parse_tokens(&mut self, iter: &mut Peekable<Iter<char>>) -> Vec<Token> {
    let mut tokens: Vec<Token> = Vec::new();
    
    while self.pos <= self.last_pos {
      self.pos += 1;
      self.line_pos += 1;
      let current = iter.next();

      tokens.push(match current {
        None => {break},
        Some(cur) => {
          match cur {
            '(' => Token{ ty: TokenType::LPar, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            ')' => Token{ ty: TokenType::RPar, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            '{' => Token{ ty: TokenType::LCurl, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            '}' => Token{ ty: TokenType::RCurl, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            '[' => Token{ ty: TokenType::LBrac, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            ']' => Token{ ty: TokenType::RBrac, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            ',' => Token{ ty: TokenType::Comma, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            ';' => Token{ ty: TokenType::Semi, line: self.line, range: Range{ start: self.pos, end: self.pos} },
            c if c.is_whitespace() => {
              if *c == '\n' {
                self.line += 1;
                self.line_pos = 0;
                Token{ ty: TokenType::Newline, line: self.line, range: Range{ start: self.pos, end: self.pos} }
              } else {
                Token{ ty: TokenType::Whitespace, line: self.line, range: Range{ start: self.pos, end: self.pos} }
              }
            },
            c if c.is_numeric() => self.lex_numbers(iter, *c),
            op if OPEARTORS.contains(&&*op.to_string()) => self.lex_operators(iter, *op),
            string if string == &'\"' => self.lex_string(iter),

            _ => Token{ ty: TokenType::Unknown, line: self.line, range: Range{ start: self.pos, end: self.pos} },
          }
        }
      })
    }
    tokens
  }

  fn lex_string(&mut self, iter: &mut Peekable<Iter<char>>) -> Token {
    let mut string = String::new();
    let start_pos = self.pos;
    let start_line = self.line;
    let mut end = false;

    while iter.peek() != None {
      if iter.peek().unwrap() == &&'\"' { 
        iter.next();
        end = true; 
        break;
      }

      if iter.peek().unwrap() == &&'\\' {
        iter.next();
        self.pos += 1;
        self.line_pos += 1;

        if iter.peek().unwrap() == &&'n' {
            string.push('\\');
            string.push('n');
            iter.next();
            self.pos += 1;
            self.line += 1;
            self.line_pos = 0;
        } else {
            string.push(*iter.next().unwrap());
            self.pos += 1;
            self.line_pos += 1;
        }
      } else {
          string.push(*iter.next().unwrap());
          self.pos += 1;
          self.line_pos += 1;
      }
    }

    if iter.peek() == None && end == false {
      panic!("")
    }

    Token{ ty: TokenType::Literal(Val::String), line: start_line, range: Range{ start: start_pos, end: self.pos} }
  }

  fn lex_operators(&mut self, iter: &mut Peekable<Iter<char>>, first: char) -> Token {
    let mut op = String::from(first);
    let start_pos = self.pos;

    while iter.peek() != None && OPEARTORS.contains(&&*iter.peek().unwrap().to_string()) {
      op.push(*iter.next().unwrap());
      self.line_pos += 1;
      self.pos += 1;
    }

    if ! OPEARTORS.contains(&&*op.to_string()) {
      panic!("")
    } else {
      Token{ ty: TokenType::new(op), line: self.line, range: Range{ start: start_pos, end: self.pos} }
    }
  }

  fn lex_numbers(&mut self, iter: &mut Peekable<Iter<char>>, first: char) -> Token {
    let mut num = String::from(first);
    let start_pos = self.pos;

    while iter.peek() != None && (iter.peek().unwrap().is_numeric() || iter.peek().unwrap() == &&'.') {
      num.push(*iter.next().unwrap());
      self.line_pos += 1;
      self.pos += 1
    }

    if num.matches('.').count() > 1 {
      panic!("")
    }

    Token{ ty: TokenType::Literal(Val::Number), line: self.line, range: Range{ start: start_pos, end: self.pos} }
  }
}